dissect/target/helpers/regutil.py

```diff
@@ -774,3 +808,126 @@ def parse_flex_value(value: str) -> ValueType:
             return struct.unpack(">Q", value)[0]
         else:
             raise NotImplementedError(f"Registry flex value type {vtype}")
+
+
+def has_glob_magic(pattern: str) -> bool:
+    """Return whether ``pattern`` contains any glob patterns
+
+    Args:
+        pattern: The string to check on glob patterns.
+
+    Returns:
+        Whether ``pattern`` contains any glob patterns.
+    """
+    return GLOB_MAGIC_REGEX.search(pattern) is not None
+
+
+def glob_split(pattern: str) -> tuple[str]:
+    """Split a key path with glob patterns on the first key path part with glob patterns
+
+    Args:
+        pattern: A key path with glob patterns to split.
+
+    Returns:
+        A tuple of two strings, where the first contains the first number of
+        key path parts (if any) which don't have a glob pattern. The second
+        contains the rest of the key path with parts containing glob patterns.
+    """
+    first_glob = GLOB_INDEX_REGEX.search(pattern)
+
+    if not first_glob:
+        return pattern, ""
+
+    pos = first_glob.start()
+    return pattern[:pos], pattern[pos:]
+
+
+def glob_ext(keycollection: KeyCollection, pattern: str) -> Iterator[KeyCollection]:
```

Comment:
Hmm, funky, that works here. I added 2 more test cases that work just like your example and they do fine?

Reply:
Maybe I am doing things wrong, but I was just trying out how I thought a real world scenario should work:

So in _os.py I added these lines:

```
@export(record=WindowsUserRecord)
    def users(self) -> Iterator[WindowsUserRecord]:
        key = "HKLM\\*\\Microsoft\\Windows NT\\CurrentVersion\\ProfileList"
        print(list(self.target.registry.glob_ext(key)))
        exit(0)
```

and I was suprised to see an empty list, but maybe I am using your globber in the wrong way. I just thought I could feed such a pattern to registry.glob_ext() and it seems to **work** except for the specified cases??? What do I miss here?

Reply:
Just a thought: does the test really create the keys we think it should make? Or are we somehow comparing "long keys" or other things without noticing? Could be an issue with the Virtual Hive thing maybe?

Reply:
It turns out `NamedKey`'s from dissect.regf include a `ROOT` part in their path. The parent of a hive has indeed a data structure that has this name, however the `NamedKey` for the root can not be retrieved using it's name, nor does the key behave "properly" (e.g. the `.path` method will give exceptions).
I'll fix this in a PR for dissect.regf.

Reply:
The `NamedKey` issues are fixed in https://github.com/fox-it/dissect.regf/pull/17

Reply:
Some other cases:

`*\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\ProfileList -> yields 0 keys`

This is understandable because the wild card is at the start, but then I expect an exception maybe?

`H?LM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\ProfileList -> yields 0 keys`

Any glob pattern in the first part causes no yield?

Is it correct that only HKLM works, not the other root keys? There seems to be support for the other keys in the code, but I get a recursion error?

Reply:
The glob pattern at the start did not work because the pattern search was modeled on how globbing for filesystems work, which always expect a path to begin with `/`. As registry key paths do not (necessarily) start with an `\` I changed that.

As for the other root keys, they should work, so I have to look into that.

Reply:
Okay, this is understandable, but maybe throw an exception if you detect such an invalid globbing pattern? Otherwise, people may see it as a potential bug (like I did) and they are going to file unnecessary bug reports. Another solution is to just add a comment of course.

Reply:
Actually, I just tested this on my fs, but ls */bin just works. But anyway, I understand the reasoning and it is fine by me, but I recommend that we make this a bit more explicit to avoid confusion.

Reply:
As for globbing on the various root keys, the short keynames (`HKCU`, `HKLM`, etc.) are not real keys in the registry. They are made up by the `registry.key()` function and replaced by the long name versions of those keys.

The `HKEY_CURRENT_USER` and `HKEY_CLASSES_ROOT` long root keys are also not physically present in the hive but mapped by `registry.key()` on other keys. So globbing on them will also not work.

Reply:
Ok, btw "*" globbing seems to work now, so fixed anyway.